File: 2022/go/day04/day04.go
package day04

import (
	"bufio"
	"fmt"
	"io"
	"strconv"
	"strings"
)

func Part1(r io.Reader) (int, error) {
	var sum int

	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		line := scanner.Text()

		fields := strings.Split(line, ",")

		elfOne, err := helper(fields[0])
		if err != nil {
			return 0, err
		}
		elfTwo, err := helper(fields[1])
		if err != nil {
			return 0, err
		}

		var elfOneCount int
		for _, v := range elfOne {
			if v == 1 {
				elfOneCount++
			}
		}

		var elfTwoCount int
		for _, v := range elfTwo {
			if v == 1 {
				elfTwoCount++
			}
		}

		small := elfOne
		large := elfTwo
		if elfTwoCount < elfOneCount {
			small = elfTwo
			large = elfOne
		}

		totalOverlap := true 
		for i := range small {
			if small[i] == 0 {
				continue
			}

			if large[i] != 1 {
				totalOverlap = false
				break
			}
		}

		if totalOverlap {
			sum++
		}
	}

	return sum, nil
}

func Part2(r io.Reader) (int, error) {
	var sum int

	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		line := scanner.Text()

		fields := strings.Split(line, ",")

		elfOne, err := helper(fields[0])
		if err != nil {
			return 0, err
		}
		elfTwo, err := helper(fields[1])
		if err != nil {
			return 0, err
		}

		for i := 0; i < 100; i++ {
			if (elfOne[i] == 1) && (elfTwo[i] == 1) {
				sum++
				break
			}
		}
	}

	return sum, nil
}

func helper(s string) ([]int, error) {
	fields := strings.Split(s, "-")

	if len(fields) < 2 {
		return nil, fmt.Errorf("invalid input %s", s)
	}

	num1, err := strconv.Atoi(fields[0])
	if err != nil {
		return nil, err
	}

	num2, err := strconv.Atoi(fields[1])
	if err != nil {
		return nil, err
	}

	res := make([]int, 100, 100)

	for num1 <= num2 {
		res[num1] = 1
		num1++
	}

	return res, nil
}

File: 2022/rust/src/day04.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

fn helper(s: &str) -> Result<[isize; 100], &'static str> {
    let fields: Vec<&str> = s.split("-").collect();

    if fields.len() < 2 {
        return Err("invalid input");
    }

    let mut num1 = fields[0].parse::<isize>().unwrap();
    let num2 = fields[1].parse::<isize>().unwrap();

    let mut res: [isize; 100] = [0; 100];

    while num1 <= num2 {
        res[num1 as usize] = 1;
        num1 = num1 + 1;
    }

    Ok(res)
}

#[allow(dead_code)]
fn part1(filename: String) -> Result<isize, &'static str> {
    let f = File::open(filename).unwrap();
    let r = BufReader::new(f);

    let mut sum: isize = 0;

    for line in r.lines() {
        let line = line.unwrap();

        let fields: Vec<&str> = line.split(",").collect();

        let elf_one = helper(fields[0]).unwrap();
        let elf_two = helper(fields[1]).unwrap();

        let mut elf_one_count: isize = 0;
        for v in elf_one.iter() {
            if *v == 1 {
                elf_one_count = elf_one_count + 1;
            }
        }

        let mut elf_two_count: isize = 0;
        for v in elf_two.iter() {
            if *v == 1 {
                elf_two_count = elf_two_count + 1;
            }
        }

        let mut small = elf_one;
        let mut large = elf_two;
        if elf_two_count < elf_one_count {
            small = elf_two;
            large = elf_one;
        }

        let mut total_overlap = true;
        for (i, _) in small.iter().enumerate() {
            if small[i] == 0 {
                continue
            }

            if large[i] != 1 {
                total_overlap = false;
                break
            }
        }

        if total_overlap {
            sum = sum + 1;
        }
    }

    return Ok(sum);
}

#[allow(dead_code)]
fn part2(filename: String) -> Result<isize, &'static str> {
    let f = File::open(filename).unwrap();
    let r = BufReader::new(f);

    let mut sum: isize = 0;

    for line in r.lines() {
        let line = line.unwrap();

        let fields: Vec<&str> = line.split(",").collect();

        let elf_one = helper(fields[0]).unwrap();
        let elf_two = helper(fields[1]).unwrap();

        for i in 0..100 {
            if elf_one[i] == 1 && elf_two[i] == 1 {
                sum = sum + 1;
                break;
            }
        }
    }

    return Ok(sum);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let result = part1(String::from("../inputs/day04.sample.txt")).unwrap();
        assert_eq!(result, 2);

        let result = part1(String::from("../inputs/day04.input.txt")).unwrap();
        assert_eq!(result, 496);
    }

    #[test]
    fn test_part2() {
        let result = part2(String::from("../inputs/day04.sample.txt")).unwrap();
        assert_eq!(result, 4);

        let result = part2(String::from("../inputs/day04.input.txt")).unwrap();
        assert_eq!(result, 847);
    }
}
